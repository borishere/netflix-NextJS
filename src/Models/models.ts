import * as yup from 'yup';

export interface Igenre {
  name: string;
  active: boolean;
}

export interface ImovieBase {
  title: string;
  overview: string;
  release_date?: string;
  poster_path: string;
  vote_average?: number;
  genres: string[];
  runtime: number;
  tagline?: string;
  vote_count?: number;
  budget?: number;
  revenue?: number;
}

export interface Imovie extends ImovieBase {
  id: number;
}

export type TNullableMovie = Imovie | null;

export interface IMoviesResponse {
  data: Imovie[];
  total: number;
  offset: number;
  limit: number;
}

export interface IMoviesState {
  sortBy: string;
  filter?: string[];
}

export interface IGetMoviesArgs {
  sortBy?: string;
  sortOrder?: 'asc' | 'desc'
  filter?: string | string[]
  search?: string,
  searchBy?: string;
  movie?: string;
}

export const validationSchema = yup.object().shape({
  title: yup.string().required(),
  release_date: yup.date(),
  vote_average: yup.number(),
  genres: yup.array().of(yup.string()).required(),
  poster_path: yup.string().url().required(),
  overview: yup.string().required(),
  runtime: yup.number().min(0).integer().required()
});