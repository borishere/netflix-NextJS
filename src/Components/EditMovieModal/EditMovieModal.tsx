import React, { FC } from 'react';
import { Formik } from 'formik';
import { Imovie, TNullableMovie, validationSchema } from '../../Models/models';
import { useEditMovieMutation } from '../../services/movies';
import { Modal, ModalProps } from '../Modal/Modal';
import { ModalForm } from '../ModalForm/ModalForm';
import styles from './EditMovieModal.module.scss';

interface EditMovieModalProps extends ModalProps {
  movie: TNullableMovie;
  updateMovie: () => void;
}

export const EditMovieModal: FC<EditMovieModalProps> = ({ movie, updateMovie, ...props }) => {
  const [editMovie, { isLoading }] = useEditMovieMutation();

  const onSubmit = (values: Imovie): void => {
    editMovie(values).then(() => {
      updateMovie();
      props.show(false);
    });
  };

  return (
    <Modal modalClass='edit-movie-modal' {...props}>
      <>
        {isLoading && <span>Editing...</span>}
        <h2 className={styles['modal-title']}>EDIT MOVIE</h2>
        {movie && (
          <Formik<Imovie>
            initialValues={movie}
            validationSchema={validationSchema}
            onSubmit={onSubmit}
          >
            <ModalForm />
          </Formik>
        )}
      </>
    </Modal>
  );
};