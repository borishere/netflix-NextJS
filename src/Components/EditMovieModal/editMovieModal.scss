.edit-movie-modal {
  .modal-content {
    max-width: 976px;
    flex: 1;
  }
}
