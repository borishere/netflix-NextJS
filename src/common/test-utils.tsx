import React, { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ModalContext } from '../Context/ModalContext';
import { Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../app/store';
import { createMemoryHistory } from 'history';
import { moviesApi } from '../services/movies';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import moviesReducer from '../app/moviesSlice';
import { configureStore } from '@reduxjs/toolkit';

const history = createMemoryHistory();


const AllTheProviders: FC = ({ children }) => {
  return (
    <Router location='/' navigator={history}>
      <Provider store={store}>
        <ModalContext.Provider value={{}}>
          {children}
        </ModalContext.Provider>
      </Provider>
    </Router>
  );
};

function renderWithStore(
  ui: ReactElement,
  {
    store = configureStore({
      reducer: {
        [moviesApi.reducerPath]: moviesApi.reducer,
        movies: moviesReducer

      },
      middleware: (getDefaultMiddleware: any) =>
        getDefaultMiddleware().concat(moviesApi.middleware)
    }),
    ...renderOptions
  } = {}
) {
  setupListeners(store.dispatch);

  function Wrapper({ children }: any) {
    const history = createMemoryHistory();

    return (
      <Provider store={store}>
        <Router location='/' navigator={history}>
          {children}
        </Router>
      </Provider>
    );
  }

  store.dispatch(moviesApi.util.resetApiState());

  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

const root = document.createElement('div');
root.id = 'root';

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, container: document.body.appendChild(root), ...options });

export * from '@testing-library/react';
export { customRender, renderWithStore };