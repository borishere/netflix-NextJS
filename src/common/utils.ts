export function addParamToExistsSearchParams(key: string, value: string): URLSearchParams {
  const params = new URLSearchParams(window.location.search);

  params.set(key, value);

  return params;
}

export function deleteParamFromExistsSearchParams(value: string): URLSearchParams {
  const params = new URLSearchParams(window.location.search);

  params.delete(value);

  return params;
}

export const BASE_API_URL: string = 'http://localhost:4000/';