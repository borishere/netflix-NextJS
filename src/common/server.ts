import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { BASE_API_URL } from './utils';
import mockData from './__mocks__/movies.json';

let movies: any[] = Array.from(mockData);

export const server = setupServer(
  rest.get(`${BASE_API_URL}movies`, (req, res, ctx) => {
    return res(ctx.json(movies.slice(0, 2)));
  }),

  rest.get(`${BASE_API_URL}movies:id`, (req, res, ctx) => {
    const { id } = req.params;
    const movie = movies.find((movie) => movie.id === id);

    return res(ctx.json(movie));
  }),

  rest.post(`${BASE_API_URL}movies`, (req, res, ctx) => {
    movies.push(req.body);

    return res(ctx.status(200));
  }),

  rest.put(`${BASE_API_URL}movies`, (req, res, ctx) => {
    const requestMovie: any = req.body;

    const movieIndex = movies.findIndex((m) => m.id === requestMovie.id);

    if (movieIndex < 0) {
      return res(ctx.status(404));
    }

    movies[movieIndex] = requestMovie;

    return res(ctx.status(200));
  }),

  rest.delete(`${BASE_API_URL}movies/:id`, (req, res, ctx) => {
    const { id } = req.params;

    movies = movies.filter((m) => m.id !== id);

    return res(ctx.status(200));
  })
);
