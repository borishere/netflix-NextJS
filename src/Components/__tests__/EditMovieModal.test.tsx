import React from 'react';
import { screen, customRender, waitFor } from '../../common/test-utils';
import { EditMovieModal } from '../EditMovieModal/EditMovieModal';
import { Imovie } from '../../Models/models';
import '@testing-library/jest-dom';

describe('EditMovieModal', () => {
  const movie: Imovie = {
    id: 76338,
    title: 'Thor: The Dark World',
    tagline: 'Delve into the darkness',
    vote_average: 6.7,
    vote_count: 6522,
    release_date: '2013-10-29',
    poster_path: 'https://image.tmdb.org/t/p/w500/bnX5PqAdQZRXSw3aX3DutDcdso5.jpg',
    overview: 'Thor fights to restore order across the cosmosâ€¦ but an ancient race led by the vengeful Malekith returns to plunge the universe back into darkness. Faced with an enemy that even Odin and Asgard cannot withstand, Thor must embark on his most perilous and personal journey yet, one that will reunite him with Jane Foster and force him to sacrifice everything to save us all.',
    budget: 170000000,
    revenue: 644571402,
    genres: [
      'documentary',
      'comedy'
    ],
    runtime: 112
  };

  test('is visible', () => {
    const { asFragment } = customRender(
      <EditMovieModal
        isShown={true}
        show={jest.fn()}
        movie={movie}
        updateMovie={jest.fn()}
      />
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('EDIT MOVIE')).toBeInTheDocument();
  });

  test('is hidden', () => {
    customRender(
      <EditMovieModal
        isShown={false}
        show={jest.fn()}
        movie={movie}
        updateMovie={jest.fn()}
      />
    );

    expect(screen.queryByText('EDIT MOVIE')).not.toBeInTheDocument();
  });

  test('Renders movie values', async () => {
    customRender(
      <EditMovieModal
        isShown={true}
        show={jest.fn()}
        movie={movie}
        updateMovie={jest.fn()}
      />
    );

    const titleInput = screen.getByLabelText('title-input');
    const overviewInput = screen.getByLabelText('overview-input');
    const ratingInput = screen.getByLabelText('rating-input');
    const runtimeInput = screen.getByLabelText('runtime-input');
    const dateInput = screen.getByLabelText('date-input');
    const urlInput = screen.getByLabelText('url-input');

    await waitFor(() => {
      expect(titleInput).toHaveValue(movie.title);
      expect(overviewInput).toHaveValue(movie.overview);
      expect(ratingInput).toHaveValue(movie.vote_average);
      expect(runtimeInput).toHaveValue(movie.runtime);
      expect(dateInput).toHaveValue(movie.release_date);
      expect(urlInput).toHaveValue(movie.poster_path);
    });
  });
});