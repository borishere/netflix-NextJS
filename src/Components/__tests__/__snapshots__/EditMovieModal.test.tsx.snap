// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditMovieModal is visible 1`] = `
<DocumentFragment>
  <div
    class="modal edit-movie-modal"
  >
    <div
      class="modal-content"
    >
      <img
        aria-label="close"
        class="modal-close-btn"
        src="test-file-stub"
      />
      <div
        class="modal-body"
      >
        <h2
          class="modal-title"
        >
          EDIT MOVIE
        </h2>
        <form
          action="#"
          class="modal-form"
        >
          <div
            class="modal-form__item-wrap"
          >
            <div
              class="modal-form__item long"
            >
              <label>
                TITLE
              </label>
              <input
                aria-label="title-input"
                class="modal-form__item-input"
                name="title"
                type="text"
                value="Thor: The Dark World"
              />
            </div>
            <div
              class="modal-form__item"
            >
              <label>
                RELEASE DATE
              </label>
              <input
                aria-label="date-input"
                class="modal-form__item-input"
                name="release_date"
                placeholder="Select Date"
                type="date"
                value="2013-10-29"
              />
            </div>
          </div>
          <div
            class="modal-form__item-wrap"
          >
            <div
              class="modal-form__item long"
            >
              <label>
                MOVIE URL
              </label>
              <input
                aria-label="url-input"
                class="modal-form__item-input"
                name="poster_path"
                placeholder="https://"
                type="text"
                value="https://image.tmdb.org/t/p/w500/bnX5PqAdQZRXSw3aX3DutDcdso5.jpg"
              />
            </div>
            <div
              class="modal-form__item"
            >
              <label>
                RATING
              </label>
              <input
                aria-label="rating-input"
                class="modal-form__item-input"
                name="vote_average"
                placeholder="7.8"
                type="number"
                value="6.7"
              />
            </div>
          </div>
          <div
            class="modal-form__item-wrap"
          >
            <div
              class="modal-form__item long"
            >
              <label>
                GENRE
              </label>
              <select
                aria-label="genres-select"
                class="modal-form__item-input"
                multiple=""
                name="genres"
                placeholder="Select Genre"
              >
                <option
                  value="Comedy"
                >
                  Comedy
                </option>
                <option
                  value="Drama"
                >
                  Drama
                </option>
                <option
                  value="Romance"
                >
                  Romance
                </option>
              </select>
            </div>
            <div
              class="modal-form__item"
            >
              <label>
                RUNTIME
              </label>
              <input
                aria-label="runtime-input"
                class="modal-form__item-input"
                name="runtime"
                placeholder="minutes"
                type="number"
                value="112"
              />
            </div>
          </div>
          <div
            class="modal-form__item"
          >
            <label>
              OVERVIEW
            </label>
            <textarea
              aria-label="overview-input"
              class="modal-form__item-input"
              name="overview"
              placeholder="movie description"
              rows="6"
            >
              Thor fights to restore order across the cosmosâ€¦ but an ancient race led by the vengeful Malekith returns to plunge the universe back into darkness. Faced with an enemy that even Odin and Asgard cannot withstand, Thor must embark on his most perilous and personal journey yet, one that will reunite him with Jane Foster and force him to sacrifice everything to save us all.
            </textarea>
          </div>
          <div
            class="modal-buttons"
          >
            <button
              class="reset-btn"
              type="reset"
            >
              RESET
            </button>
            <button
              class="submit-btn"
              type="submit"
            >
              SUBMIT
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
